backend/cn/check.ml

```diff
@@ -530,26 +530,22 @@ let rec check_pexpr (pe : BT.t mu_pexpr) (k : IT.t -> unit m) : unit m =
        | ty -> fail (fun _ -> {loc; msg = Mismatch {has = BT.pp ty; expect = !^"comparable type"}})
      in
      begin match op with
-     | OpAdd ->
-      let@ () = WellTyped.ensure_base_type loc ~expect: (bt_of_pexpr pe1) (bt_of_pexpr pe2) in
-         check_pexpr pe1 (fun v1 ->
-         check_pexpr pe2 (fun v2 ->
-         k (add_ (v1, v2) loc)))
-     | OpSub ->
-         let@ () = WellTyped.ensure_base_type loc ~expect: (bt_of_pexpr pe1) (bt_of_pexpr pe2) in
-         check_pexpr pe1 (fun v1 ->
-         check_pexpr pe2 (fun v2 ->
-         k (sub_ (v1, v2) loc)))
-     | OpMul ->
-         let@ () = WellTyped.ensure_base_type loc ~expect: (bt_of_pexpr pe1) (bt_of_pexpr pe2) in
-         check_pexpr pe1 (fun v1 ->
-         check_pexpr pe2 (fun v2 ->
-         k (mul_ (v1, v2) loc)))
      | OpDiv ->
          let@ () = WellTyped.ensure_base_type loc ~expect: (bt_of_pexpr pe1) (bt_of_pexpr pe2) in
          check_pexpr pe1 (fun v1 ->
          check_pexpr pe2 (fun v2 ->
-         k (div_ (v1, v2) loc)))
+          let@ provable = provable loc in
+          let v2_bt = IT.bt v2 in
+          let zero = match v2_bt with
+          | BT.Bits (Signed, size) -> int_lit_ 0 (BT.Bits (Signed, size)) loc
+          | BT.Bits (Unsigned, size) -> int_lit_ 0 (BT.Bits (Unsigned, size)) loc
+          | BT.Integer -> int_lit_ 0 (BT.Integer) loc
+          | _ -> failwith "expecting a divisble type"
+          in
```

Comment:
Please use `ensure_bits` to check bt of pexpr_pe2 is a bit vector. Using failwith will result in bad error experience.